### ISSUE TITLE
Spellchecking for markdown areas

### ISSUE THREAD

**Issue by ellisonbg** [jupyter/jupyterlab#846]
Users should be able to spellcheck narrative text in markdown areas, this would cover all instances of markdown, including the text editor and markdown cells. Some link to Doug Blanks work on this in the classic notebook:

https://github.com/Calysto/notebook-extensions/tree/master/calysto/spell-check
https://www.npmjs.com/package/typo-js

**Comment by blink1073** [comment]
I'd say this should belong as an extension, since it is inherently locale-specific.

**Comment by dsblank** [comment]
It is indeed locale specific (like lots of software) but it should be at least trivial to include in a deployed service (note, or jupyterhub). Having a writing tool (like the notebook) without spelling-checking is a fundamental flaw. Perhaps this really a deploy/install issue?

**Comment by jasongrout** [comment]
+1 to having it be an extension. I think this falls into the same category as linters and other things that would like to annotate/decorate text, though. I think it would make sense to explore having a more generic framework for an extension to annotate or decorate text.

For reference, there is a spell check extension for visual studio code: https://github.com/Microsoft/vscode-spell-check

**Comment by drorata** [comment]
Is there some update on this issue? The normal Jupyter notebook has this functionality implemented using an [extension](http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/spellchecker/README.html). I find it a very important feature which would hold me from adopting JupyterLab.

**Comment by chriswbishop** [comment]
+1 to extension

**Comment by cbarrick** [comment]
Is it possible to delegate to the browser's builtin spell check?

That would solve the locale issue, since the user's browser would presumably be configured to the proper locale, and it would reduce the maintenance burden.

Since we use CodeMirror for the text boxes, the solution might need to be upstream. They have an existing issue (codemirror/CodeMirror#1017) that has been closed because it's not possible to accomplish in their current architecture. It seems that browsers only spell check in `contenteditable` elements like `<input>`.

Someone _could_ try to replace CodeMirror with something `contenteditable`. A daunting task for sure...

**Comment by aaronzs** [comment]
I just find this [jupyterlab_spellchecker](https://github.com/ijmbarr/jupyterlab_spellchecker).
Give a try! 😄

**Comment by shivendra90** [comment]
+ 1 for spell checking. @aaronzs That extension is good but it has issues when creating new cells. Works only while editing.

**Comment by ellisonbg** [comment]
I would love to see this extension eventually migrated into the core jupyterlab.

**Comment by ellisonbg** [comment]
The above linked VSCode extension is deprecated in favor of this one:

https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

**Comment by krassowski** [comment]
Is there still interest in bringing spellchecker into core?

**Comment by jasongrout** [comment]
Yes, I would love to see a spellchecker in core. I think for me, it's more of a question of technical implementation and size, and what those tradeoffs are.

**Comment by stackfuel** [comment]
Just dropping an idea here: It would already be helpful, if one could temporarily replace the whole Code Mirror stuff of one cell with a simple textarea, perhaps via a menu item in the top menu bar somewhere.

Or second idea: Even to simply render the Code Mirror contained text separately again inside a textarea below the cell, removable later again. Such a thing could be realized by an extension as well though.

Both of these ideas do not require to completely get rid of Code Mirror. Code Mirror in general however, adds loads of bloat in terms of the DOM. So if JupyterLab got rid of that at some point, I would count that as a plus. It might not be easy to find a simpler less bloaty text editor though, which offers the same kind of functionality.

**Comment by krassowski** [comment]
FYI In turn of events, the extension moved to https://github.com/jupyterlab-contrib/jupyterlab_spellchecker where we will continue developing it to make the size and implementation more bearable, with a potential outlook for integration into the core in the future. The recent success it porting it to federated extension, but more work ahead.
